### PR TITLE
AF: volume→surface cross-attention for volume closure

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Volume prediction is harder because volume points are spatially distant from the geometry surface where aerodynamic information originates. The current model treats surface and volume points identically through the same transformer trunk — volume points must implicitly learn to leverage surface information via self-attention. But in a 2-layer transformer with mixed surface+volume tokens, this implicit path may be too weak.

We add an explicit **cross-attention layer** after the transformer trunk where **volume points attend to surface point representations**. This gives volume predictions direct access to surface-level aerodynamic information — physically motivated because volume flow fields in incompressible flow are determined by surface boundary conditions (potential flow theory).

Surface points bypass the cross-attention and go directly to the output MLP (their predictions are already strong). Only volume points benefit from the enriched representation.

## Instructions

All changes in `target/icml2026/train.py`. Champion architecture and hyperparameters unchanged — only add the cross-attention post-processing.

**Step 1: Add cross-attention module to the model**

```python
class VolumeSurfaceCrossAttention(nn.Module):
    def __init__(self, hidden_dim, num_heads=4):
        super().__init__()
        self.cross_attn = nn.MultiheadAttention(
            embed_dim=hidden_dim,
            num_heads=num_heads,
            batch_first=True,
        )
        self.norm = nn.LayerNorm(hidden_dim)

    def forward(self, volume_features, surface_features):
        # volume_features: [N_vol, hidden_dim] — queries
        # surface_features: [N_surf, hidden_dim] — keys/values
        # Add batch dimension for nn.MultiheadAttention
        vol = volume_features.unsqueeze(0)   # [1, N_vol, D]
        surf = surface_features.unsqueeze(0)  # [1, N_surf, D]
        attn_out, _ = self.cross_attn(vol, surf, surf)  # [1, N_vol, D]
        # Residual connection + LayerNorm
        out = self.norm(volume_features + attn_out.squeeze(0))
        return out  # [N_vol, hidden_dim]
```

**Step 2: Integrate after transformer trunk**

In `__init__`:
```python
self.vol_surf_cross_attn = VolumeSurfaceCrossAttention(
    hidden_dim=256, num_heads=4
)
```

In `forward()`, after the transformer trunk produces features for all points:
```python
# all_features: [N_total, hidden_dim] from transformer trunk
# Separate surface and volume features (check how existing code tracks point types)
surface_features = all_features[surface_mask]  # [N_surf, D]
volume_features = all_features[volume_mask]    # [N_vol, D]

# Enrich volume with surface context
volume_features = self.vol_surf_cross_attn(volume_features, surface_features)

# Recombine for output MLP (or apply separate heads)
all_features[volume_mask] = volume_features
output = self.output_mlp(all_features)
```

Check how the data pipeline separates surface/volume points — look at how `vol_loss` vs `surface_loss` are computed.

**Step 3: Run with `--wandb_group af-vol-cross-attention`**

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0 \
  --wandb_group af-vol-cross-attention
```

**Report:** both surface_mse and vol_mse at best checkpoint, epoch counts, throughput (the cross-attention is one extra attention layer on volume points only — expect modest overhead, should still complete 600+ epochs).

## Baseline

- **val surface_mse:** 0.000296 at epoch 704
- **val vol_mse:** 0.002039 at epoch 743
- **W&B run:** PR #3135, EMA=0.999 + vol-weight=10x champion
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0`

## Success criteria

- **Primary:** vol_mse < 0.002039 without surface_mse regression above 0.000320
- **Stretch:** vol_mse < 0.0017 (SpiderSolver benchmark target)
- Surface is a hard constraint — even large vol gains are rejected if surface regresses significantly

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### Summary

The volume→surface cross-attention hypothesis **did not meet success criteria**. Both surface_mse and vol_mse remained well above baselines despite 518 epochs of training (6h timeout).

### Metrics

| Metric | Baseline | This Run | Ratio | Target |
|--------|----------|----------|-------|--------|
| **val surface_mse** | 0.000296 (ep704) | **0.001007** (ep406) | 3.4x worse | < 0.000320 |
| **val vol_mse** | 0.002039 (ep743) | **0.003032** (ep515) | 1.5x worse | < 0.002039 |
| test surface_mse | — | 0.001063 | — | — |
| test vol_mse | — | 0.003578 | — | — |

#### Per-component test breakdown

| Component | surface_mse | volume_mse |
|-----------|------------|------------|
| Ux | 0.000046 | 0.006237 |
| Uy | 0.000002 | 0.004223 |
| nut | 0.000002 | 0.003249 |
| p | 0.004201 | 0.000601 |

#### Convergence at cosine troughs (T_max=50)

| Trough | best surface_mse | vol_mse | surface gap | vol gap |
|--------|-----------------|---------|-------------|---------|
| ep50 | 0.004600 | 0.039496 | 15.5x | 19.4x |
| ep150 | 0.004600 | 0.012294 | 15.5x | 6.0x |
| ep250 | 0.001854 | 0.006380 | 6.3x | 3.1x |
| ep350 | 0.001372 | 0.004523 | 4.6x | 2.2x |
| ep450 | 0.001007 | 0.003502 | 3.4x | 1.7x |
| ep500 | 0.001007 | 0.003221 | 3.4x | 1.6x |

### Training Details

- **Command:**
  ```bash
  CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full \
    --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 \
    --weight-decay 1e-2 --enable-fourier --model-layers 2 \
    --model-hidden-dim 256 --model-heads 4 --epochs 999 \
    --ema-decay 0.999 --volume-loss-weight 10.0 --vol-cross-attn \
    --agent casca --wandb-name "casca/af-vol-cross-attn" \
    --wandb-group af-vol-cross-attention
  ```
- **W&B run:** `4u3jvnbd`
- **Epochs:** 518 (360-min timeout)
- **Throughput:** ~1.4 ep/min (0.7 min/epoch) — baseline does ~1.5 ep/min, so ~7% overhead from cross-attention
- **Best checkpoint:** epoch 406 (val surface_mse), epoch 515 (val vol_mse)
- **EMA decay:** 0.999

### What happened

**The cross-attention did not help — it made both metrics significantly worse.**

1. **Surface regression is severe (3.4x).** The wrapper model wraps the base SenpaiTransolver, so surface predictions pass through unchanged. The regression likely comes from the VolumeCrossAttentionWrapper adding a correction head that uses the same backbone hidden states. The wrapper applies `outputs["volume_preds"] + correction`, and the zero-initialized correction head starts neutral but then changes the optimization landscape for shared backbone parameters. The backbone is no longer optimizing solely for surface accuracy — it now also has gradients flowing back through the cross-attention and correction head, which compete with the surface loss signal.

2. **Volume also worse (1.5x at best val).** The cross-attention enrichment path (volume→surface attention → linear correction) adds a residual correction to the base volume predictions. Despite zero-initialization for stable start, the correction head appears to introduce noise rather than improvement. The attention over ~7k surface keys from ~170k volume queries may be too diffuse — each volume point gets a weighted average of all surface representations, which is essentially a global mean rather than local geometric information.

3. **Convergence is much slower.** Baseline reaches best surface at ep704 vs this run at ep406 — but the run only completed 518 epochs. Even projecting to ep700+, the convergence trajectory (still at 3.4x gap at ep500) suggests it would not close the gap.

### Why it failed

The core issue is that the cross-attention wrapper modifies the optimization dynamics of the shared backbone. Even though surface predictions are computed identically (the wrapper only adds to volume_preds), the gradients from the volume loss now flow through both the original volume head AND the new cross-attention path. This changes how the backbone weights are updated, degrading surface predictions as a side effect.

The physical intuition (volume fields are determined by surface boundary conditions) is sound, but applying it as a post-hoc wrapper on a pre-trained architecture recipe doesn't work because:
- The backbone already implicitly mixes surface/volume info via self-attention
- Adding a second attention path creates gradient competition
- The correction is global (all surface → each volume point) rather than geometrically local

### Suggested follow-ups

1. **Integrate cross-attention within the backbone** rather than as a wrapper — add it as an extra block between existing transformer layers, so the model learns to use it from the start rather than fighting existing optimization dynamics.
2. **Geometric locality** — use distance-weighted or kNN-gated cross-attention so volume points attend to nearby surface points, not all of them.
3. **Freeze surface path** — if using a wrapper approach, freeze the backbone and only train the cross-attention + correction head, so surface predictions are guaranteed not to regress.
4. **This direction may not be worth pursuing** — the 2-layer backbone with 96 slices already has implicit surface↔volume mixing. The marginal benefit of explicit cross-attention may be too small to justify the complexity and optimization difficulty.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3282 because ADVISOR: Closing — vol→surface cross-attention surface=0.001007 (3.4x worse), vol=0.003032 (1.5x worse) after 518 epochs. Architectural modification fundamentally failed. Casca reassigned..
